doc/configuration.rst
.. _configuration_settings:

Configuration Settings
======================

Configuration can be done either by using a configuration file or with environnement variable.

Except stated otherwise, the rule for environnement variable names is ``QGSRV_<SECTION>_<KEY>`` all in uppercase.


Common Configuration Options
=============================





.. _SERVER_HTTP_PORT:

SERVER_HTTP_PORT
----------------

Port to listen to

:Type: int
:Default: 8080
:Section: server
:Key: port
:Env: QGSRV_SERVER_HTTP_PORT



.. _SERVER_INTERFACES:

SERVER_INTERFACES
-----------------

Interfaces to listen to


:Type: string
:Default: 0.0.0.0
:Section: server
:Key: interfaces
:Env: QGSRV_SERVER_INTERFACES



.. _SERVER_OUTPUTFILE_AS_REFERENCE:

SERVER_OUTPUTFILE_AS_REFERENCE
------------------------------

Select if output files are returned by reference for deferred download or in response body as 
complex output.


:Type: boolean
:Default: no
:Section: server
:Key: outputfile_as_reference
:Env: QGSRV_SERVER_OUTPUTFILE_AS_REFERENCE



.. _SERVER_MAXINPUTSIZE:

SERVER_MAXINPUTSIZE
-------------------

Max input file fetched as remote reference

:Type: size
:Default: 100m
:Section: server
:Key: maxinputsize
:Env: QGSRV_SERVER_MAXINPUTSIZE



.. _SERVER_MAXBUFFERSIZE:

SERVER_MAXBUFFERSIZE
--------------------

Max request buffer size.

:Type: size
:Default: 1m
:Section: server
:Key: maxbuffersize
:Env: QGSRV_SERVER_MAXBUFFERSIZE



.. _SERVER_WORKDIR:

SERVER_WORKDIR
--------------

Parent working directory where processes are executed. Eache processes will create
a working directiry for storing results files and logs. 
The default value use the `gettempdir()` function.


:Type: path
:Default: System defaults
:Section: server
:Key: workdir
:Env: QGSRV_SERVER_WORKDIR



.. _SERVER_PARALLELPROCESSES:

SERVER_PARALLELPROCESSES
------------------------

The number of parallel processes runninc `execute` requests. Extra processes will be queued.


:Type: int
:Default: 1
:Section: server
:Key: parallelprocesses
:Env: QGSRV_SERVER_PARALLELPROCESSES



.. _SERVER_PROCESSLIFECYCLE:

SERVER_PROCESSLIFECYCLE
-----------------------

Maximal number of executions that can run in the same worker before beeing recreating
the worker.


:Type: int
:Default: 1
:Section: server
:Key: processlifecycle
:Env: QGSRV_SERVER_PROCESSLIFECYCLE



.. _SERVER_MAXQUEUESIZE:

SERVER_MAXQUEUESIZE
-------------------

Maximal number of waiting tasks - extra tasks will return a 509 in synchronous execution.


:Type: int
:Default: 100
:Section: server
:Key: maxqueuesize
:Env: QGSRV_SERVER_MAXQUEUESIZE



.. _SERVER_RESPONSE_TIMEOUT:

SERVER_RESPONSE_TIMEOUT
-----------------------

Timeout for tasks execution in seconds. Task running longer that this time will be aborted and
a timeout error is retourned.


:Type: int
:Default: 1800
:Section: server
:Key: response_timeout
:Env: QGSRV_SERVER_RESPONSE_TIMEOUT



.. _SERVER_RESPONSE_EXPIRATION:

SERVER_RESPONSE_EXPIRATION
--------------------------

Response expiration in seconds. After that delay from the tasks's end, data (working directory and status)
for that task will be deleted.


:Type: int
:Default: 86400
:Section: server
:Key: response_expiration
:Env: QGSRV_SERVER_RESPONSE_EXPIRATION



.. _SERVER_WMS_SERVICE_URL:

SERVER_WMS_SERVICE_URL
----------------------

The url for the service used to retrieve results as WMS/WFS references.
Usually this will correspond to a Qgis server serving OWS services from results projects.



:Type: string
:Default: Request host url
:Section: server
:Key: wms_service_url
:Env: QGSRV_SERVER_WMS_SERVICE_URL



.. _SERVER_RESULTS_MAP_URI:

SERVER_RESULTS_MAP_URI
----------------------

Base uri used for the MAP argument in WMS/WFS response references.
Define a base URI to use for 'MAP' arguments in WMS/WFS responses, this uri may
corresponds to an 'alias in py-qgis-server <https://py-qgis-server.readthedocs.io/en/latest/schemes.html#scheme-aliases>' _.



:Type: string
:Default: wps_result_map_uri
:Section: server
:Key: wms_service_url
:Env: QGSRV_SERVER_RESULTS_MAP_URI



.. _SERVER_CROSS_ORIGIN:

SERVER_CROSS_ORIGIN
-------------------

Allows any origin for CORS. If set to 'no', allow only CORS for the 'Origin'
header.


:Type: boolean
:Section: server
:Key: cross_origin
:Env: QGSRV_SERVER_CROSS_ORIGIN



.. _SERVER_RESTARTMON:

SERVER_RESTARTMON
-----------------

The file to watch for restarting workers. When the modified date of the file is changed, 
the processing providers are reloaded.


:Type: path
:Section: server
:Key: restartmon
:Env: QGSRV_SERVER_RESTARTMON



.. _LOGGING_LEVEL:

LOGGING_LEVEL
-------------

Set the logging level

:Type: ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']
:Default: DEBUG
:Section: logging
:Key: level
:Env: QGSRV_LOGGING_LEVEL



.. _REDIS_HOST:

REDIS_HOST
----------

Redis storage backend host


:Type: string
:Default: localhost
:Section: logstorage:redis
:Key: host
:Env: QGSRV_REDIS_HOST



.. _REDIS_PORT:

REDIS_PORT
----------

Redis storage backend port


:Type: string
:Default: 6379
:Section: logstorage:redis
:Key: port
:Env: QGSRV_REDIS_PORT



.. _REDIS_DBNUM:

REDIS_DBNUM
-----------

Redis storage backend database index


:Type: string
:Section: logstorage:redis
:Key: dbnum
:Env: QGSRV_REDIS_DBNUM



.. _REDIS_PREFIX:

REDIS_PREFIX
------------

Redis storage backend key prefix.


:Type: string
:Default: pyqgiswps
:Section: logstorage:redis
:Key: prefix
:Env: QGSRV_REDIS_PREFIX



.. _CACHE_SIZE:

CACHE_SIZE
----------

The maximal number of Qgis projects held in cache. The cache strategy is LRU.


:Type: int
:Default: 10
:Section: projects.cache
:Key: size
:Env: QGSRV_CACHE_SIZE



.. _CACHE_ROOTDIR:

CACHE_ROOTDIR
-------------

The directory location for Qgis project files.


:Type: path
:Section: projects.cache
:Key: rootdir
:Env: QGSRV_CACHE_ROOTDIR



.. _CACHE_STRICT_CHECK:

CACHE_STRICT_CHECK
------------------

Activate strict checking of project layers. When enabled, Qgis projects
with invalid layers will be dismissed and an 'Unprocessable Entity' (422) HTTP error
will be issued.


:Type: boolean
:Default: yes
:Section: projects.cache
:Key: strict_check
:Env: QGSRV_CACHE_STRICT_CHECK



.. _PROCESSING_PROVIDERS_MODULE_PATH:

PROCESSING_PROVIDERS_MODULE_PATH
--------------------------------

Path to Qgis processing providers modules

:Type: path
:Section: processing
:Key: providers_module_path
:Env: QGSRV_PROCESSING_PROVIDERS_MODULE_PATH



.. _PROCESSING_EXPOSED_PROVIDERS:

PROCESSING_EXPOSED_PROVIDERS
----------------------------

Path to Qgis processing providers modules

:Type: list
:Default: script,model
:Section: processing
:Key: exposed_providers
:Env: QGSRV_PROCESSING_EXPOSED_PROVIDERS



.. _PROCESSING_ACCESSPOLICY:

PROCESSING_ACCESSPOLICY
-----------------------

Path to the access policy configuration file

:Type: path
:Default: PROCESSING_PROVIDERS_MODULE_PATH/accesspolicy.yml
:Section: processing
:Key: accesspolicy
:Env: QGSRV_PROCESSING_ACCESSPOLICY




pyqgiswps/config.py
#
# Copyright 2018 3liz
# Author: David Marteau
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Original parts are Copyright 2016 OSGeo Foundation,            
# represented by PyWPS Project Steering Committee,               
# and released under MIT license.                                
# Please consult PYWPS_LICENCE.txt for details
# Original Author: Calin Ciociu                                           
#

"""
Reads the PyWPS configuration file
"""

import sys
import os
import tempfile
import functools

import configparser

from typing import Any
from pyqgisservercontrib.core import componentmanager


CONFIG = configparser.ConfigParser(interpolation=configparser.ExtendedInterpolation())
CONFIG.optionxform = lambda opt: opt

def _log( *args ):
    print( *args, file=sys.stderr, flush=True)


def load_configuration():
    """Load PyWPS configuration from configuration file.

    :param cfgfile: path to the configuration file
    :param cfgdefault: default configuration dict
    :param cfgdict: configuration dict (override file and default)
    """

    CONFIG.clear()

    _log('loading configuration')
    getenv = os.getenv

    #
    # Server
    #

    CONFIG.add_section('server')
    CONFIG.set('server', 'encoding', 'utf-8')
    CONFIG.set('server', 'language', 'en-US')
    CONFIG.set('server', 'url', '{host_url}')

    CONFIG.set('server', 'port'          , getenv('QGSWPS_SERVER_HTTP_PORT', '8080'))
    CONFIG.set('server', 'interfaces'    , getenv('QGSWPS_SERVER_INTERFACES', '0.0.0.0'))

    # Select if output files are returned by reference or in response body
    CONFIG.set('server', 'outputfile_as_reference', getenv('QGSWPS_SERVER_OUTPUTFILE_AS_REFERENCE', 'yes'))
    # Max input file fetched as remote reference
    CONFIG.set('server', 'maxinputsize'         , getenv('QGSWPS_SERVER_MAXINPUTSIZE' ,'100m'))
    # Max request size 
    CONFIG.set('server', 'maxbuffersize'        , getenv('QGSWPS_SERVER_MAXBUFFERSIZE','1m'))
    # Storage url for retrieving reference files
    CONFIG.set('server', 'store_url'            , '{host_url}store/{uuid}/{file}?service=WPS')
    # URL for retrieving status
    CONFIG.set('server', 'status_url'           , '{host_url}ows/?service=WPS&request=GetResults&uuid={uuid}') 
    # Path to workdir where process are executed"
    CONFIG.set('server', 'workdir'              , getenv('QGSWPS_SERVER_WORKDIR',tempfile.gettempdir()))
    # Configure host if the server is behind a proxy
    CONFIG.set('server', 'host_proxy'           , getenv('QGSWPS_SERVER_HOST_PROXY',''))
    # Number of parallel processes that are allowed to run algorithms
    CONFIG.set('server', 'parallelprocesses'    , getenv('QGSWPS_SERVER_PARALLELPROCESSES','1'))
    # Maximal number of executions can run in the same worker before beeing restarted
    CONFIG.set('server', 'processlifecycle'     , getenv('QGSWPS_SERVER_PROCESSLIFECYCLE','1'))
    # Maximal number of waiting tasks - extra tasks will return a 509 in synchronous execution
    CONFIG.set('server', 'maxqueuesize'         , getenv('QGSWPS_SERVER_MAXQUEUESIZE','100'))
    # Timeout for tasks execution
    CONFIG.set('server', 'response_timeout'     , getenv('QGSWPS_SERVER_RESPONSE_TIMEOUT','1800'))
    # Expiration time in Redis cache for task responses
    CONFIG.set('server', 'response_expiration'  , getenv('QGSWPS_SERVER_RESPONSE_EXPIRATION','86400'))
    # Base url used for return WMS references (Qgis projects holding layers created by WPS tasks)
    CONFIG.set('server', 'wms_service_url'      , getenv('QGSWPS_SERVER_WMS_SERVICE_URL','${wps.request:host_url}'))
    # Base uri used for the MAP argument in WMS references
    CONFIG.set('server', 'wps_result_map_uri'   , getenv('QGSWPS_SERVER_RESULTS_MAP_URI','wps-results:'))
    # Full URL used for returning  WPS references
    CONFIG.set('server', 'wms_response_uri'     , '${wms_service_url}?MAP={map_url}&service=WMS&request=GetCapabilities')
    # Cleanup interval in seconds
    CONFIG.set('server', 'cleanup_interval'     ,'600')
    # Download API expiration ttl for a download URL
    CONFIG.set('server', 'download_ttl'         , getenv('QGSWPS_DOWNLOAD_TTL','30'))
    # Enable middleware filters from extensions
    CONFIG.set('server', 'enable_filters'       , getenv('QGSWPS_SERVER_ENABLE_FILTERS', 'yes'))
    # Path for pre-installed extensions
    CONFIG.set('server', 'sys_config_path'      , getenv('QGSWPS_SYS_CONFIG_PATH'      , '/usr/share/qgis_wps'))
    CONFIG.set('server', 'cross_origin'         , getenv('QGSWPS_SERVER_CROSS_ORIGIN' , 'yes'))
    CONFIG.set('server', 'restartmon'           , getenv('QGSWPS_SERVER_RESTARTMON' , ''))

    #
    # Logging
    #

    CONFIG.add_section('logging')
    CONFIG.set('logging', 'level', getenv('QGSWPS_LOGLEVEL','INFO'))

    #
    # Log storage
    #

    CONFIG.add_section('logstorage:redis')
    CONFIG.set('logstorage:redis', 'host'  , getenv('QGSWPS_REDIS_HOST'  ,'localhost'))
    CONFIG.set('logstorage:redis', 'port'  , getenv('QGSWPS_REDIS_PORT'  ,'6379'))
    CONFIG.set('logstorage:redis', 'dbnum' , getenv('QGSWPS_REDIS_DBNUM' ,'0')) 
    CONFIG.set('logstorage:redis', 'prefix', getenv('QGSWPS_REDIS_PREFIX','pyqgiswps')) 

    #
    # Projects cache
    #

    CONFIG.add_section('projects.cache')
    # Maximun number of projects in cache
    CONFIG.set('projects.cache', 'size'    , getenv('QGSWPS_CACHE_SIZE','10' ))
    # Root directory for projects file
    CONFIG.set('projects.cache', 'rootdir' , getenv('QGSWPS_CACHE_ROOTDIR',''))
    # Ensure that loaded project is valid before loading in cache
    CONFIG.set('projects.cache', 'strict_check' , getenv('QGSWPS_CACHE_STRICT_CHECK','yes'))

    CONFIG.add_section('projects.schemes')

    #
    # Processing
    #

    CONFIG.add_section('processing')
    CONFIG.set('processing', 'providers_module_path', getenv('QGSWPS_PROCESSING_PROVIDERS_MODULE_PATH',''))
    CONFIG.set('processing', 'exposed_providers'    , getenv('QGSWPS_PROCESSING_EXPOSED_PROVIDERS' ,'script,model'))
    CONFIG.set('processing', 'accesspolicy'         , getenv('QGSWPS_PROCESSING_ACCESSPOLICY' ,'${providers_module_path}/accesspolicy.yml'))
    CONFIG.set('processing', 'default_module_path'  , '${server:sys_config_path}/processing')

    #
    # Qgis folders settings
    # 
    # Enable to define search folder list with globbing patterns
    # to be expanded in Qgis settings
    #
    CONFIG.add_section('qgis.settings.folders')
    CONFIG.set('qgis.settings.folders', 'Processing/Configuration/SCRIPTS_FOLDERS', '${processing:providers_module_path}/scripts')
    CONFIG.set('qgis.settings.folders', 'Processing/Configuration/MODELS_FOLDER'  , '${processing:providers_module_path}/models')


    # Qgis settings
    #
    # Add other Qgis settings
    #
    # XXX Not functional yet
    CONFIG.add_section('qgis.settings')

    #
    # Qgis projects
    #
    CONFIG.add_section('qgis.projects')
    # Set this to force the the advertised WMS url in the Qgis projects
    # created by pyqgiswps
    CONFIG.set('qgis.projects', 'wmsurl', getenv('QGSWPS_ADVERTISED_WMSURL','${server:wms_service_url}'))

    #
    # Metadata
    #

    CONFIG.add_section('metadata:main')
    CONFIG.set('metadata:main', 'identification_title', 'Py-Qgis-WPS Processing Service')
    CONFIG.set('metadata:main', 'identification_abstract', 'Py-Qgis-WPS is an implementation of the Web Processing Service standard from the Open Geospatial Consortium. Py-Qgis-WPS is written in Python.')  # noqa
    CONFIG.set('metadata:main', 'identification_keywords', 'Py-Qgis-WPS,WPS,OGC,QGIS,processing')
    CONFIG.set('metadata:main', 'identification_keywords_type', 'theme')
    CONFIG.set('metadata:main', 'identification_fees', 'NONE')
    CONFIG.set('metadata:main', 'identification_accessconstraints', 'NONE')
    CONFIG.set('metadata:main', 'provider_name', 'Organization Name')
    CONFIG.set('metadata:main', 'provider_url', 'https://github.com/3liz/py-qgis-wps')
    CONFIG.set('metadata:main', 'contact_name', 'Lastname, Firstname')
    CONFIG.set('metadata:main', 'contact_address', 'Mailing Address')
    CONFIG.set('metadata:main', 'contact_email', 'Email Address')
    CONFIG.set('metadata:main', 'contact_url', 'Contact URL')
    CONFIG.set('metadata:main', 'contact_role', 'pointOfContact')

    #
    # Used at runtime to set contextual configuration
    #
    CONFIG.add_section('wps.request')


def read_config_dict( userdict ):
    """ Read configuration from dictionary

        Will override previous settings
    """
    CONFIG.read_dict( userdict )


def write_config( fh ):
    """ Dump configuration to file
    """
    CONFIG.write(fh)


def read_config_file( cfgfile ):
    """ Read configuration from file
    """
    cfgfile = os.path.abspath(cfgfile)
    with open(cfgfile, mode='rt') as fp:
        CONFIG.read_file(fp)
    _log('Configuration file <%s> loaded' % cfgfile)


def config_to_dict():
    """ Convert actual configuration to dictionary
    """
    def _items(p):
        for k in p.keys():
            try:
                yield k,p[k]
            except configparser.InterpolationMissingOptionError:
                yield k,p.get(k,raw=True)

    return { s: dict(_items(p)) for s,p in CONFIG.items() }


def validate_config_path(confname, confid, optional=False):
    """ Validate directory path
    """
    confvalue = CONFIG.get(confname,confid,fallback='')

    if not confvalue and optional:
        return

    confvalue = os.path.normpath(confvalue)
    if not os.path.isdir(confvalue):
        _log('ERROR: server->%s configuration value %s is not directory' % (confid, confvalue))
        raise ValueError(confvalue)

    if not os.path.isabs(confvalue):
        _log('ERROR: server->%s configuration value %s is not absolute path' % (confid, confvalue))
        raise ValueError(confvalue)

    CONFIG.set(confname, confid, confvalue)


def get_size_bytes(size):
    """Get real size of given obeject

    """
    size = size.lower()

    import re

    units = re.compile("[gmkb].*")
    newsize = float(re.sub(units, '', size))

    if size.find("g") > -1:
        newsize *= 1024 * 1024 * 1024
    elif size.find("m") > -1:
        newsize *= 1024 * 1024
    elif size.find("k") > -1:
        newsize *= 1024
    else:
        newsize *= 1
    return newsize


#
# Published services
#

NO_DEFAULT=object()

# Chars that are to be replaced in env variable name
ENV_REPLACE_CHARS=':.-@#$%&*'

@componentmanager.register_factory('@3liz.org/config-service;1')
class ConfigService:
    """ Act as a proxy
    """

    def __init__(self):
        self.allow_env = True

    def __get_impl( self, _get_fun, section:str, option:str, fallback:Any = NO_DEFAULT ) -> Any:
        """
        """
        value = _get_fun(section, option, fallback=NO_DEFAULT)
        if value is NO_DEFAULT:
            # Look in environment
            # Note that the section must exists
            if self.allow_env:
                varname  = 'QGSWPS_%s_%s' % (section.upper(), option.upper())
                varname  = functools.reduce( lambda s,c: s.replace(c,'_'), ENV_REPLACE_CHARS, varname)
                varvalue = os.getenv(varname)
                if varvalue is not None:
                    CONFIG.set(section, option, varvalue)
                # Let config parser translate the value for us
                value = _get_fun(section, option, fallback=fallback)
            else:
                value = fallback
        if value is NO_DEFAULT:
            raise KeyError('[%s] %s' % (section,option))
        return value


    get        = functools.partialmethod(__get_impl,CONFIG.get)
    getint     = functools.partialmethod(__get_impl,CONFIG.getint)
    getboolean = functools.partialmethod(__get_impl,CONFIG.getboolean)
    getfloat   = functools.partialmethod(__get_impl,CONFIG.getfloat)

    def items( self, section: str ):
        return CONFIG.items(section)

    def __getitem__(self, section):
        return CONFIG[section]

    def __contains__(self, section):
        return section in CONFIG

    def set( self, section:str, option:str, value: Any ) -> None:
        CONFIG.set( section, option, value )

    def add_section( self, sectionname: str ) -> None:
        # We do not care if the section already exists
        try:
            CONFIG.add_section( sectionname )
        except configparser.DuplicateSectionError:
            pass 


confservice = ConfigService()


